File: frontend/src/lib/businessDays.ts
/** Feriados nacionais fixos brasileiros: [mês, dia, nome] */
const FIXED_HOLIDAYS: [number, number, string][] = [
  [1,  1,  "Ano Novo"],
  [4,  21, "Tiradentes"],
  [5,  1,  "Dia do Trabalho"],
  [9,  7,  "Independência"],
  [10, 12, "N. Sra. Aparecida"],
  [11, 2,  "Finados"],
  [11, 15, "Proclamação da República"],
  [12, 25, "Natal"],
];

export const WEEKDAYS_PT = ["Dom", "Seg", "Ter", "Qua", "Qui", "Sex", "Sáb"];
export const MONTHS_PT = [
  "", "Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho",
  "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro",
];

export function isBusinessDay(d: Date): boolean {
  const wd = d.getDay(); // 0=Dom, 6=Sáb
  if (wd === 0 || wd === 6) return false;
  const m = d.getMonth() + 1;
  const day = d.getDate();
  return !FIXED_HOLIDAYS.some(([hm, hd]) => hm === m && hd === day);
}

export function getSkipReason(d: Date): string | null {
  const wd = d.getDay();
  if (wd === 6) return "Sábado";
  if (wd === 0) return "Domingo";
  const m = d.getMonth() + 1;
  const day = d.getDate();
  const holiday = FIXED_HOLIDAYS.find(([hm, hd]) => hm === m && hd === day);
  return holiday ? `Feriado: ${holiday[2]}` : null;
}

export function getNextBusinessDay(d: Date): Date {
  const result = new Date(d);
  while (!isBusinessDay(result)) {
    result.setDate(result.getDate() + 1);
  }
  return result;
}

export interface PaymentInfo {
  targetDay: number;
  originalDate: Date;
  paymentDate: Date;
  isAdjusted: boolean;
  skipReason: string | null;
  weekday: string;
}

export function getSalaryPaymentInfo(
  year: number,
  month: number,
  targetDay: number
): PaymentInfo {
  // Último dia do mês
  const maxDay = new Date(year, month, 0).getDate();
  const actualDay = Math.min(targetDay, maxDay);
  const original = new Date(year, month - 1, actualDay);
  const payment = getNextBusinessDay(original);
  const isAdjusted = original.getTime() !== payment.getTime();

  return {
    targetDay,
    originalDate: original,
    paymentDate: payment,
    isAdjusted,
    skipReason: isAdjusted ? getSkipReason(original) : null,
    weekday: WEEKDAYS_PT[payment.getDay()],
  };
}

File: backend/business_days.py
"""
Utilitários para cálculo de dias úteis brasileiros.
Considera fins de semana e feriados nacionais fixos.
"""
import calendar
from datetime import date, timedelta
from typing import Optional

MONTHS_PT = [
    "", "Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho",
    "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro",
]

WEEKDAYS_PT = ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"]

# Feriados nacionais fixos (mês, dia): nome
FIXED_HOLIDAYS: dict[tuple[int, int], str] = {
    (1, 1):   "Ano Novo",
    (4, 21):  "Tiradentes",
    (5, 1):   "Dia do Trabalho",
    (9, 7):   "Independência do Brasil",
    (10, 12): "N. Sra. Aparecida",
    (11, 2):  "Finados",
    (11, 15): "Proclamação da República",
    (12, 25): "Natal",
}


def is_business_day(d: date) -> bool:
    """Retorna True se a data for um dia útil (seg–sex, não feriado nacional)."""
    if d.weekday() >= 5:  # 5=Sábado, 6=Domingo
        return False
    if (d.month, d.day) in FIXED_HOLIDAYS:
        return False
    return True


def get_next_business_day(d: date) -> date:
    """Retorna o próximo dia útil (o mesmo dia se já for útil)."""
    while not is_business_day(d):
        d += timedelta(days=1)
    return d


def get_skip_reason(d: date) -> Optional[str]:
    """Retorna o motivo pelo qual a data não é útil, ou None se for útil."""
    wd = d.weekday()
    if wd == 5:
        return "Sábado"
    if wd == 6:
        return "Domingo"
    return FIXED_HOLIDAYS.get((d.month, d.day))


def get_salary_payment_info(year: int, month: int, target_day: int) -> dict:
    """
    Calcula a data real de pagamento do salário para o mês/ano dado.
    Se o dia alvo cair em fim de semana ou feriado, avança para o próximo dia útil.

    Returns:
        dict com target_day, original_date, payment_date, payment_day,
        is_adjusted, skip_reason, weekday
    """
    max_day = calendar.monthrange(year, month)[1]
    actual_day = min(target_day, max_day)
    original_date = date(year, month, actual_day)
    adjusted_date = get_next_business_day(original_date)
    is_adjusted = original_date != adjusted_date
    reason = get_skip_reason(original_date) if is_adjusted else None

    return {
        "target_day": target_day,
        "original_date": original_date.isoformat(),
        "payment_date": adjusted_date.isoformat(),
        "payment_day": adjusted_date.day,
        "is_adjusted": is_adjusted,
        "skip_reason": reason,
        "weekday": WEEKDAYS_PT[adjusted_date.weekday()],
    }
